### ISSUE TITLE
UnsupportedOperation: read  -- AtomicLocalFile, flatten, doc / code mismatch

### ISSUE THREAD

**Issue by computercolin** [spotify/luigi#2598]
First off, thank you for your fantastic library! Really, its great!

Now, issue: When I do the following:
  - Create `Task` with `output()` that returns `AtomicLocalFile`
  - Execute luigi scheduler

I would expect tasks to be scheduled, run, everything is cool.

Instead: `UnsupportedOperation: read`

Cause: during scheduling, as tasks are checked for _completion_, Luigi runs `task.flatten` on targets. When passed `AtomicLocalFile`, `task.flatten` ends up here:

```python
    try:
        # if iterable
        iterator = iter(struct)
    except TypeError:
        return [struct]
    for result in iterator:
        flat += flatten(result)
    return flat
```
And treats `AtomicLocalFile(io.BufferedWriter)` as an iterator. This goes badly, `BufferedWriter` gives an iterator, but when used, iterator throws the `UnsupportedOperation: read` because its a writer, not a reader (.. sort of raises the question of why BufferedReader implements \_\_iter\_\_ at all).

There are many many ways to fix this. Add `__iter__(): raise NotHappeningError()` to `AtomicLocalFile` / `local_target.atomic_file` . Change `task.flatten` to check for BufferedIO children. Wrap BufferedWriter in the class, instead of extending it...

I'm new, so I don't totally know if these classes are intended for something else, but it seems Luigi developed, and `AtomicLocalFile` didn't catch up. Or possibly just the docs didn't catch up.

Docs say `AtomicLocalFile`: Abstract class to create a **Target** that creates a temporary file in the local filesystem before moving it to its final destination.

Is `AtomicLocalFile` meant to be a "Target," or more of a convenience auto-move-bufferedwriter-opener? I see several subclasses in `contrib` that subclass AtomicLocalFile but none of them override \_\_iter\_\_ (so none can be returned by `output()`)

Since I don't have my sea legs yet just a check: large point of Targets is they can be returned by `output()`, right? Or is there another use?
It seems, for batch file processing where you implement all/most pieces, it is more *DRY* to output() targets of the desired type and use them directly, instead of using a generic "data container" Target, and having to instantiate the specific target you want. But.. I'm new, so ???.

Note tested  on python 3.5 and 3.6

**Comment by dlstadther** [comment]
Hi @computercolin ! Thanks for the thorough issue submission!

Now that i've had some time to read through this and digest the perceived problem, I'll try to provide a response.

`AtomicLocalFile` is not a luigi class with which I have personal experience with. However, the wording of its docstring, its usage in [local_target.py](https://github.com/spotify/luigi/blob/master/luigi/local_target.py#L34-L43), and its [PR](https://github.com/spotify/luigi/pull/769) lead me to believe that it was provided as a helper middle-man class for the creation of a file-type Target. Unfortunately, there are no unittests which utilize `AtomicLocalFile` directly.

In general, [`output()` expects to return some form of a `Target`](https://luigi.readthedocs.io/en/latest/tasks.html#task-output) - be that a `LocalTarget`, `S3Target`,  `RedshiftTarget`, etc.


Hope this helps!

**Comment by computercolin** [comment]
Thanks @dlstadther 
I was surprised I was able to find the error I did within very limited use of Luigi -- so I figured I was doing something strange. As you say, `AtomicLocalFile` is a helper, and not really meant to be be used directly as a target (even if implementing its 1 abstract method is very easy). I toyed around with `AtomicLocalFile` because it seemed to offer something that LocalTarget did not. But behold, I am very wrong, `LocalTarget` uses AtomicLocalFile internally itself and isn't missing anything!

I don't think there is a "bug" here, at least not in code. But I think the docstring for `AtomicLocalFile` needs some changes.

> Current: Abstract class to create a Target that creates  a temporary file in the local filesystem before  moving it to its final destination.

Sure, it does create a "bytes-type-**target**-you-can-write-to" (probably what author meant)... but when I read that docs, from `luigi/target.py`, only way I interpret "Target" is as: "creates a luigi-**Target** that creates a temporary file in the local filesystem...

maybe something like

>Idea: Abstract helper class for use in Target implementations. Use in place of open() or BufferedWriter. Opens temp file for writing (bytes mode) and triggers `move_to_final_destination` on normal-close. ... subclasses must implement `move_to_final_destination`...

Might seem a bit specific, but the whole package docs are in readthedocs https://luigi.readthedocs.io/en/stable/api/luigi.html and its pretty easy to stumble into luigi.target and think "ah great, AtomicLocal... lets use it!"
